## Practice Exercise 2

In this assignment, you will try to find some interesting insights into a few movies released between 1916 and 2016, using Python. You will have to download a movie dataset, write Python code to explore the data, gain insights into the movies, actors, directors, and collections, and submit the code.

#### Some tips before starting the assignment

1. Identify the task to be performed correctly, and only then proceed to write the required code. Don’t perform any incorrect analysis or look for information that isn’t required for the assignment.
2. In some cases, the variable names have already been assigned, and you just need to write code against them. In other cases, the names to be given are mentioned in the instructions. We strongly advise you to use the mentioned names only.
3. Always keep inspecting your data frame after you have performed a particular set of operations.
4. There are some checkpoints given in the IPython notebook provided. They're just useful pieces of information you can use to check if the result you have obtained after performing a particular task is correct or not.
5. Note that you will be asked to refer to documentation for solving some of the questions. That is done on purpose for you to learn new commands and also how to use the documentation.

In [1]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

### Task 1: Reading and Inspection

**Subtask 1.1: Import and read**

Import and read the movie database. Store it in a variable called `movies`.

In [2]:
# Write your code for importing the csv file here
movies = pd.read_csv('6_Movies.csv')
movies

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4     Color     Andrew Stanton                   462.0     132.0   
...     ...                ...                     ...       ...   
3848  Color      Shane Carruth                   143.0      77.0   
3849  Color   Neill Dela Llana                    35.0      80.0   
3850  Color   Robert Rodriguez                    56.0      81.0   
3851  Color       Edward Burns                    14.0      95.0   
3852  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
0                         0.0                   855.0    Joel David Moore   
1                       563.0                  1000.0       Orlando Bloom   
2                         0.0                   161.0        Rory Kinnear   
3                     22000.0                 23000.0      Christian Bale   
4                       475.0                   530.0     Samantha Morton   
...                       ...                     ...                 ...   
3848                    291.0                     8.0      David Sullivan   
3849                      0.0                     0.0     Edgar Tancangco   
3850                      0.0                     6.0     Peter Marquardt   
3851                      0.0                   133.0  Caitlin FitzGerald   
3852                     16.0                    16.0    Brian Herzlinger   

      actor_1_facebook_likes        gross  \
0                     1000.0  760505847.0   
1                    40000.0  309404152.0   
2                    11000.0  200074175.0   
3                    27000.0  448130642.0   
4                      640.0   73058679.0   
...                      ...          ...   
3848                   291.0     424760.0   
3849                     0.0      70071.0   
3850                   121.0    2040920.0   
3851                   296.0       4584.0   
3852                    86.0      85222.0   

                                   genres  ... num_user_for_reviews language  \
0         Action|Adventure|Fantasy|Sci-Fi  ...               3054.0  English   
1                Action|Adventure|Fantasy  ...               1238.0  English   
2               Action|Adventure|Thriller  ...                994.0  English   
3                         Action|Thriller  ...               2701.0  English   
4                 Action|Adventure|Sci-Fi  ...                738.0  English   
...                                   ...  ...                  ...      ...   
3848                Drama|Sci-Fi|Thriller  ...                371.0  English   
3849                             Thriller  ...                 35.0  English   
3850  Action|Crime|Drama|Romance|Thriller  ...                130.0  Spanish   
3851                         Comedy|Drama  ...                 14.0  English   
3852                          Documentary  ...                 84.0  English   

          country  content_rating       budget  title_year  \
0             USA           PG-13  237000000.0      2009.0   
1             USA           PG-13  300000000.0      2007.0   
2              UK           PG-13  245000000.0      2015.0   
3             USA           PG-13  250000000.0      2012.0   
4             USA           PG-13  263700000.0      2012.0   
...           ...             ...          ...         ...   
3848          USA           PG-13       7000.0      2004.0   
3849  Philippines       Not Rated       7000.0      2005.0   
3850          USA               R       7000.0      1992.0   
3851          USA       Not Rated       9000.0      2011.0   
3852          USA              PG       1100.0      2004.0   

     actor_2_face

**Subtask 1.2: Inspect the dataframe**

Inspect the dataframe's columns, shapes, variable types etc.

In [3]:
# Write your code for inspection here
movies.shape

(3853, 28)

In [4]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3853 entries, 0 to 3852
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      3851 non-null   object 
 1   director_name              3853 non-null   object 
 2   num_critic_for_reviews     3852 non-null   float64
 3   duration                   3852 non-null   float64
 4   director_facebook_likes    3853 non-null   float64
 5   actor_3_facebook_likes     3847 non-null   float64
 6   actor_2_name               3852 non-null   object 
 7   actor_1_facebook_likes     3853 non-null   float64
 8   gross                      3853 non-null   float64
 9   genres                     3853 non-null   object 
 10  actor_1_name               3853 non-null   object 
 11  movie_title                3853 non-null   object 
 12  num_voted_users            3853 non-null   int64  
 13  cast_total_facebook_likes  3853 non-null   int64

In [5]:
movies.describe

<bound method NDFrame.describe of       color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4     Color     Andrew Stanton                   462.0     132.0   
...     ...                ...                     ...       ...   
3848  Color      Shane Carruth                   143.0      77.0   
3849  Color   Neill Dela Llana                    35.0      80.0   
3850  Color   Robert Rodriguez                    56.0      81.0   
3851  Color       Edward Burns                    14.0      95.0   
3852  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
0                         0.0                   855.0    Joel David Moor

#### <font color='red'>Question 1: How many rows and columns are present in the dataframe? </font>
-  <font color='red'>(3821, 26)</font>
-  <font color='red'>(3879, 28)</font>
-  <font color='red'>(3853, 28)</font>
-  <font color='red'>(3866, 26)</font>

In [6]:
## Write your code here
movies.shape

(3853, 28)

#### <font color='red'>Question 2: How many columns have null values present in them? Try writing a code for this instead of counting them manually.</font>

-  <font color='red'>3</font>
-  <font color='red'>6</font>
-  <font color='red'>9</font>
-  <font color='red'>12</font>

In [7]:
## Write your code here
null_column_count = movies.columns[movies.isna().any()]

print(len(null_column_count))

12


### Task 2: Cleaning the Data

**Subtask 2.1: Drop unecessary columns**

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So it is advised to drop the following columns.
-  color
-  director_facebook_likes
-  actor_1_facebook_likes
-  actor_2_facebook_likes
-  actor_3_facebook_likes
-  actor_2_name
-  cast_total_facebook_likes
-  actor_3_name
-  duration
-  facenumber_in_poster
-  content_rating
-  country
-  movie_imdb_link
-  aspect_ratio
-  plot_keywords

In [8]:
# Check the 'drop' function in the Pandas library - dataframe.drop(list_of_unnecessary_columns, axis = )
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations

movies.drop(columns=[
    'color', 
    'director_facebook_likes', 
    'actor_1_facebook_likes', 
    'actor_2_facebook_likes', 
    'actor_3_facebook_likes', 
    'actor_2_name', 
    'cast_total_facebook_likes', 
    'actor_3_name', 
    'duration', 
    'facenumber_in_poster', 
    'content_rating', 
    'country', 
    'movie_imdb_link', 
    'aspect_ratio', 
    'plot_keywords'], 
    inplace = True)

#### <font color='red'>Question 3: What is the count of columns in the new dataframe? </font>
-  <font color='red'>10</font>
-  <font color='red'>13</font>
-  <font color='red'>15</font>
-  <font color='red'>17</font>

In [9]:
## Write your code here
movies.shape

(3853, 13)

#### Count of columns = 13

**Subtask 2.2: Inspect Null values**

As you have seen above, there are null values in multiple columns of the dataframe 'movies'. Find out the percentage of null values in each column of the dataframe 'movies'. 

In [10]:
# Write you code here
percent_missing = movies.isnull().sum() * 100 / len(movies)
print(percent_missing)

director_name             0.000000
num_critic_for_reviews    0.025954
gross                     0.000000
genres                    0.000000
actor_1_name              0.000000
movie_title               0.000000
num_voted_users           0.000000
num_user_for_reviews      0.000000
language                  0.077861
budget                    0.000000
title_year                0.000000
imdb_score                0.000000
movie_facebook_likes      0.000000
dtype: float64


#### <font color='red'>Question 4: Which column has the highest percentage of null values? </font>
-  <font color='red'>language</font>
-  <font color='red'>genres</font>
-  <font color='red'>num_critic_for_reviews</font>
-  <font color='red'>imdb_score</font>

In [11]:
## Write your code here
df = pd.DataFrame({
    'col_name': movies.columns,
    'percent_missing': percent_missing
})

df.sort_values('percent_missing', ascending = False)

col_name  percent_missing
language                              language         0.077861
num_critic_for_reviews  num_critic_for_reviews         0.025954
director_name                    director_name         0.000000
gross                                    gross         0.000000
genres                                  genres         0.000000
actor_1_name                      actor_1_name         0.000000
movie_title                        movie_title         0.000000
num_voted_users                num_voted_users         0.000000
num_user_for_reviews      num_user_for_reviews         0.000000
budget                                  budget         0.000000
title_year                          title_year         0.000000
imdb_score                          imdb_score         0.000000
movie_facebook_likes      movie_facebook_likes         0.000000

**Subtask 2.3: Fill NaN values**

You might notice that the `language` column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with `'English'`.

In [12]:
# Write your code for filling the NaN values in the 'language' column here
movies['language'].fillna('English', inplace = True)

#### <font color='red'>Question 5: What is the count of movies made in English language after replacing the NaN values with English? </font>
-  <font color='red'>3670</font>
-  <font color='red'>3674</font>
-  <font color='red'>3668</font>
-  <font color='red'>3672</font>

In [13]:
## Write your code here
print(movies['language'].value_counts()['English'])

3674


### Task 3: Data Analysis

**Subtask 3.1: Change the unit of columns**

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

In [14]:
## Write your code here
movies['budget'] = movies['budget'] / 1000000
movies['gross'] = movies['gross'] / 1000000

In [15]:
movies[['gross', 'budget']]

gross    budget
0     760.505847  237.0000
1     309.404152  300.0000
2     200.074175  245.0000
3     448.130642  250.0000
4      73.058679  263.7000
...          ...       ...
3848    0.424760    0.0070
3849    0.070071    0.0070
3850    2.040920    0.0070
3851    0.004584    0.0090
3852    0.085222    0.0011

[3853 rows x 2 columns]

**Subtask 3.2: Find the movies with highest profit**

   1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
   2. Sort the dataframe using the `profit` column as reference. (Find which command can be used here to sort entries from the documentation)
   3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [16]:
# Write your code for creating the profit column here
movies['profit'] = movies['gross'] - movies['budget']

In [17]:
# Write your code for sorting the dataframe here
movies.sort_values('profit', ascending = False, inplace = True)

In [18]:
#top10 = # Write your code to get the top 10 profiting movies here
movies.head(10)

director_name  num_critic_for_reviews       gross  \
0         James Cameron                   723.0  760.505847   
28      Colin Trevorrow                   644.0  652.177271   
25        James Cameron                   315.0  658.672302   
2704       George Lucas                   282.0  460.935665   
2748   Steven Spielberg                   215.0  434.949459   
16          Joss Whedon                   703.0  623.279547   
482        Roger Allers                   186.0  422.783777   
230        George Lucas                   320.0  474.544677   
64    Christopher Nolan                   645.0  533.316061   
419           Gary Ross                   673.0  407.999255   

                                        genres         actor_1_name  \
0              Action|Adventure|Fantasy|Sci-Fi          CCH Pounder   
28            Action|Adventure|Sci-Fi|Thriller  Bryce Dallas Howard   
25                               Drama|Romance    Leonardo DiCaprio   
2704           Action|Adventure|Fantasy|Sci-Fi        Harrison Ford   
2748                             Family|Sci-Fi         Henry Thomas   
16                     Action|Adventure|Sci-Fi      Chris Hemsworth   
482   Adventure|Animation|Drama|Family|Musical    Matthew Broderick   
230            Action|Adventure|Fantasy|Sci-Fi      Natalie Portman   
64                 Action|Crime|Drama|Thriller       Christian Bale   
419            Adventure|Drama|Sci-Fi|Thriller    Jennifer Lawrence   

                                     movie_title  num_voted_users  \
0                                        Avatar            886204   
28                               Jurassic World            418214   
25                                      Titanic            793059   
2704         Star Wars: Episode IV - A New Hope            911097   
2748                 E.T. the Extra-Terrestrial            281842   
16                                 The Avengers            995415   
482                               The Lion King            644348   
230   Star Wars: Episode I - The Phantom Menace            534658   
64                              The Dark Knight           1676169   
419                            The Hunger Games            701607   

      num_user_for_reviews language  budget  title_year  imdb_score  \
0                   3054.0  English   237.0      2009.0         7.9   
28                  1290.0  English   150.0      2015.0         7.0   
25                  2528.0  English   200.0      1997.0         7.7   
2704                1470.0  English    11.0      1977.0         8.7   
2748                 515.0  English    10.5      1982.0         7.9   
16                  1722.0  English   220.0      2012.0         8.1   
482                  656.0  English    45.0      1994.0         8.5   
230                 3597.0  English   115.0      1999.0         6.5   
64                  4667.0  English   185.0      2008.0         9.0   
419                 1959.0  English    78.0      2012.0         7.3   

      movie_facebook_likes      profit  
0                    33000  523.505847  
28                  150000  502.177271  
25                   26000  458.672302  
2704                 33000  449.935665  
2748                 34000  424.449459  
16                  123000  403.279547  
482                  17000  377.783777  
230                  13000  359.544677  
64                   37000  348.316061  
419                 140000  329.999255

**Checkpoint:** You might spot two movies directed by `James Cameron` in the list.

#### <font color='red'>Question 6: Which movie is ranked 5th from the top in the list obtained? </font>
-  <font color='red'>E.T. the Extra-Terrestrial</font>
-  <font color='red'>The Avengers</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='red'>Titanic</font>

In [19]:
## Write your code here
movies.iloc[4]

director_name                        Steven Spielberg
num_critic_for_reviews                          215.0
gross                                      434.949459
genres                                  Family|Sci-Fi
actor_1_name                             Henry Thomas
movie_title               E.T. the Extra-Terrestrial 
num_voted_users                                281842
num_user_for_reviews                            515.0
language                                      English
budget                                           10.5
title_year                                     1982.0
imdb_score                                        7.9
movie_facebook_likes                            34000
profit                                     424.449459
Name: 2748, dtype: object

**Subtask 3.3: Find IMDb Top 200**

Create a new dataframe `IMDb_Top_200` and store the top 200 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`). Also make sure that for all of these movies, the `num_voted_users` is greater than 25,000. 

Also add a `Rank` column containing the values 1 to 200 indicating the ranks of the corresponding films.

In [20]:
# Write your code for extracting the top 200 movies as per the IMDb score here. Make sure that you store it in a new dataframe 
# and name that dataframe as 'IMDb_Top_200'
IMDb_Top_200 = movies[movies.num_voted_users > 25000]
IMDb_Top_200 = IMDb_Top_200.sort_values('imdb_score', ascending = False)
IMDb_Top_200 = IMDb_Top_200.head(200)

In [43]:
IMDb_Top_200.shape

(200, 16)

In [44]:
IMDb_Top_200['Rank'] = IMDb_Top_200['imdb_score'].rank()

In [45]:
IMDb_Top_200.sort_values('Rank', ascending = False, inplace = True)
IMDb_Top_200

director_name  num_critic_for_reviews       gross  \
1795        Frank Darabont                   199.0   28.341469   
3016  Francis Ford Coppola                   208.0  134.821952   
2543  Francis Ford Coppola                   149.0   57.300000   
64       Christopher Nolan                   645.0  533.316061   
325          Peter Jackson                   328.0  377.019252   
...                    ...                     ...         ...   
3196      Jean-Marc Vallée                   471.0   27.296514   
1900           Spike Jonze                   574.0   25.556065   
335              Brad Bird                   283.0  261.437578   
2488           Danny Boyle                   418.0  141.319195   
3437         Walter Salles                    71.0    5.595428   

                                 genres         actor_1_name  \
1795                        Crime|Drama       Morgan Freeman   
3016                        Crime|Drama            Al Pacino   
2543                        Crime|Drama       Robert De Niro   
64          Action|Crime|Drama|Thriller       Christian Bale   
325      Action|Adventure|Drama|Fantasy        Orlando Bloom   
...                                 ...                  ...   
3196                    Biography|Drama  Matthew McConaughey   
1900               Drama|Romance|Sci-Fi   Scarlett Johansson   
335   Action|Adventure|Animation|Family         Holly Hunter   
2488                      Drama|Romance          Anil Kapoor   
3437                              Drama  Fernanda Montenegro   

                                         movie_title  num_voted_users  \
1795                       The Shawshank Redemption           1689764   
3016                                  The Godfather           1155770   
2543                         The Godfather: Part II            790926   
64                                  The Dark Knight           1676169   
325   The Lord of the Rings: The Return of the King           1215718   
...                                              ...              ...   
3196                             Dallas Buyers Club            326494   
1900                                            Her            355126   
335                                 The Incredibles            479166   
2488                            Slumdog Millionaire            641997   
3437                                Central Station             28951   

      num_user_for_reviews    language  budget  title_year  imdb_score  \
1795                4144.0     English    25.0      1994.0         9.3   
3016                2238.0     English     6.0      1972.0         9.2   
2543                 650.0     English    13.0      1974.0         9.0   
64                  4667.0     English   185.0      2008.0         9.0   
325                 3189.0     English    94.0      2003.0         8.9   
...                    ...         ...     ...         ...         ...   
3196                 340.0     English     5.0      2013.0         8.0   
1900                 748.0     English    23.0      2013.0         8.0   
335                  815.0     English    92.0      2004.0         8.0   
2488                1017.0     English    15.0      2008.0         8.0   
3437                 257.0  Portuguese     2.9      1998.0         8.0   

      movie_facebook_likes      profit   Rank category_rating  
1795                108000    3.341469  200.0        9 to 9.5  
3016                 43000  128.821952  199.0        9 to 9.5  
2543                 14000   44.300000  197.5        9 to 9.5  
64                   37000  348.316061  197.5        9 to 9.5  
325                  16000  283.019252  194.5        8.5 to 9  
...                    ...         ...    ...             ...  
3196                 65000   22.296514   24.5        8 to 8.5  
1900                148000    2.556065   24.5        8 to 8.5  
335                      0  169.437578   24.5        8 to 8.5  
2488                 13000  126.319195   24.5

#### <font color='red'>Question 7: Suppose movies are divided into 5 buckets based on the IMDb ratings: </font>
-  <font color='red'>7.5 to 8</font>
-  <font color='red'>8 to 8.5</font>
-  <font color='red'>8.5 to 9</font>
-  <font color='red'>9 to 9.5</font>
-  <font color='red'>9.5 to 10</font>

<font color = 'red'> Which bucket holds the maximum number of movies from *IMDb_Top_200*? </font>

In [40]:
## Write your code here
def get_imdb_rating(rating):
    if rating >= 7.5 and rating < 8:
        return '7.5 to 8'
    elif rating >= 8 and rating < 8.5:
        return '8 to 8.5'
    elif rating >= 8.5 and rating < 9:
        return '8.5 to 9'
    elif rating >= 9 and rating < 9.5:
        return '9 to 9.5'
    else:
        return '9.5 to 10'

In [41]:
IMDb_Top_200['category_rating'] = IMDb_Top_200['imdb_score'].apply(get_imdb_rating)

In [42]:
test = IMDb_Top_200['category_rating'].value_counts()
test

8 to 8.5    154
8.5 to 9     42
9 to 9.5      4
Name: category_rating, dtype: int64

**Subtask 3.4: Find the critic-favorite and audience-favorite actors**

   1. Create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the `actor_1_name` column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
   2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
   3. Group the combined dataframe using the `actor_1_name` column.
   4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [27]:
# Write your code for creating three new dataframes here
#Meryl_Streep = # Include all movies in which Meryl_Streep is the lead
Meryl_Streep = movies[movies.actor_1_name == 'Meryl Streep']
Meryl_Streep.shape

(11, 14)

In [28]:
#Leo_Caprio = # Include all movies in which Leo_Caprio is the lead
Leo_Caprio = movies[movies.actor_1_name == 'Leonardo DiCaprio']
Leo_Caprio.shape

(21, 14)

In [29]:
#Brad_Pitt = # Include all movies in which Brad_Pitt is the lead
Brad_Pitt = movies[movies.actor_1_name == 'Brad Pitt']
Brad_Pitt.shape

(17, 14)

In [30]:
# Write your code for combining the three dataframes here
# Combined = 
Combined = pd.concat([Meryl_Streep, Leo_Caprio, Brad_Pitt])
Combined.shape

(49, 14)

In [31]:
# Write your code for grouping the combined dataframe here
Combined.groupby('actor_1_name').sum()

num_critic_for_reviews        gross  num_voted_users  \
actor_1_name                                                              
Brad Pitt                          4165.0  1127.464628          4820925   
Leonardo DiCaprio                  6934.0  2640.581774          9281136   
Meryl Streep                       1996.0   659.117002           809001   

                   num_user_for_reviews  budget  title_year  imdb_score  \
actor_1_name                                                              
Brad Pitt                       12620.0  1178.0     34074.0       121.6   
Leonardo DiCaprio               19204.0  1731.5     42115.0       157.4   
Meryl Streep                     3269.0   379.0     22039.0        74.2   

                   movie_facebook_likes      profit  
actor_1_name                                         
Brad Pitt                        238880  -50.535372  
Leonardo DiCaprio               1096000  909.081774  
Meryl Streep                      32275  280.117002

In [32]:
# Write the code for finding the mean of critic reviews and audience reviews here
Combined.groupby('actor_1_name').mean()

num_critic_for_reviews       gross  num_voted_users  \
actor_1_name                                                             
Brad Pitt                      245.000000   66.321449    283583.823529   
Leonardo DiCaprio              330.190476  125.741989    441958.857143   
Meryl Streep                   181.454545   59.919727     73545.545455   

                   num_user_for_reviews     budget   title_year  imdb_score  \
actor_1_name                                                                  
Brad Pitt                    742.352941  69.294118  2004.352941    7.152941   
Leonardo DiCaprio            914.476190  82.452381  2005.476190    7.495238   
Meryl Streep                 297.181818  34.454545  2003.545455    6.745455   

                   movie_facebook_likes     profit  
actor_1_name                                        
Brad Pitt                  14051.764706  -2.972669  
Leonardo DiCaprio          52190.476190  43.289608  
Meryl Streep                2934.090909  25.465182

#### <font color='red'>Question 8: Which actor is highest rated among the three actors according to the user reviews? </font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

In [33]:
## Write your code here
combined_user_review = Combined.groupby('actor_1_name').mean().sort_values('num_user_for_reviews', ascending = False)
combined_user_review

num_critic_for_reviews       gross  num_voted_users  \
actor_1_name                                                             
Leonardo DiCaprio              330.190476  125.741989    441958.857143   
Brad Pitt                      245.000000   66.321449    283583.823529   
Meryl Streep                   181.454545   59.919727     73545.545455   

                   num_user_for_reviews     budget   title_year  imdb_score  \
actor_1_name                                                                  
Leonardo DiCaprio            914.476190  82.452381  2005.476190    7.495238   
Brad Pitt                    742.352941  69.294118  2004.352941    7.152941   
Meryl Streep                 297.181818  34.454545  2003.545455    6.745455   

                   movie_facebook_likes     profit  
actor_1_name                                        
Leonardo DiCaprio          52190.476190  43.289608  
Brad Pitt                  14051.764706  -2.972669  
Meryl Streep                2934.090909  25.465182

#### <font color='red'>Question 9: Which actor is highest rated among the three actors according to the critics?</font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

In [34]:
## Write your code here
combined_critic_review = Combined.groupby('actor_1_name').mean().sort_values('num_critic_for_reviews', ascending = False)
combined_critic_review

num_critic_for_reviews       gross  num_voted_users  \
actor_1_name                                                             
Leonardo DiCaprio              330.190476  125.741989    441958.857143   
Brad Pitt                      245.000000   66.321449    283583.823529   
Meryl Streep                   181.454545   59.919727     73545.545455   

                   num_user_for_reviews     budget   title_year  imdb_score  \
actor_1_name                                                                  
Leonardo DiCaprio            914.476190  82.452381  2005.476190    7.495238   
Brad Pitt                    742.352941  69.294118  2004.352941    7.152941   
Meryl Streep                 297.181818  34.454545  2003.545455    6.745455   

                   movie_facebook_likes     profit  
actor_1_name                                        
Leonardo DiCaprio          52190.476190  43.289608  
Brad Pitt                  14051.764706  -2.972669  
Meryl Streep                2934.090909  25.465182

#### <font color='red'>Question 10: Which movie starting with the letter A, has the 3rd highest profit?</font>
-  <font color='red'>Austin Powers: The Spy Who Shagged Me</font>
-  <font color='red'>Avatar</font>
-  <font color='red'>Avengers: Age of Ultron</font>

In [35]:
## Write your code here
movies_with_A = movies[movies['movie_title'].str.startswith('A')].sort_values('profit', ascending = False)
movies_with_A.iloc[2]

director_name                          Joss Whedon
num_critic_for_reviews                       635.0
gross                                   458.991599
genres                     Action|Adventure|Sci-Fi
actor_1_name                       Chris Hemsworth
movie_title               Avengers: Age of Ultron 
num_voted_users                             462669
num_user_for_reviews                        1117.0
language                                   English
budget                                       250.0
title_year                                  2015.0
imdb_score                                     7.5
movie_facebook_likes                        118000
profit                                  208.991599
Name: 7, dtype: object

#### <font color='red'>Question 11: Among the movies starting with the letter A, who has the highest imdb score?</font>
-  <font color='red'>American History X</font>
-  <font color='red'>Apocalypse Now</font>
-  <font color='red'>Avengers: Age of Ultron</font>

In [36]:
## Write your code here
movies_with_A = movies[movies['movie_title'].str.startswith('A')].sort_values('imdb_score', ascending = False)
movies_with_A.iloc[0]

director_name                       Tony Kaye
num_critic_for_reviews                  162.0
gross                                6.712241
genres                            Crime|Drama
actor_1_name                     Ethan Suplee
movie_title               American History X 
num_voted_users                        782437
num_user_for_reviews                   1420.0
language                              English
budget                                    7.5
title_year                             1998.0
imdb_score                                8.6
movie_facebook_likes                    35000
profit                              -0.787759
Name: 2826, dtype: object

#### <font color='red'>Question 12: Among the movies that '*Joss Whedon*' has directed, which of them made the highest profit?</font>
-  <font color='red'>Serenity</font>
-  <font color='red'>The Avengers</font>
-  <font color='red'>Avengers: Age of Ultron</font>

In [37]:
## Write your code here
jw_movies = movies[movies.director_name == 'Joss Whedon'].sort_values('profit', ascending = False)
jw_movies.iloc[0]

director_name                         Joss Whedon
num_critic_for_reviews                      703.0
gross                                  623.279547
genres                    Action|Adventure|Sci-Fi
actor_1_name                      Chris Hemsworth
movie_title                         The Avengers 
num_voted_users                            995415
num_user_for_reviews                       1722.0
language                                  English
budget                                      220.0
title_year                                 2012.0
imdb_score                                    8.1
movie_facebook_likes                       123000
profit                                 403.279547
Name: 16, dtype: object

#### <font color='red'>Question 13: Among the movies that were released in the year '*2007*' & '*2008*', which of them had the 2nd highest imdb score?</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='red'>WALL·E</font>
-  <font color='red'>No End in Sight</font>
-  <font color='red'>Into the Wild</font>

In [38]:
## Write your code here
movies_07 = movies[movies.title_year == 2007]
movies_08 = movies[movies.title_year == 2008]
movies_07_08 = pd.concat([movies_07, movies_08])

In [39]:
movies_07_08.sort_values('imdb_score', ascending = False).iloc[1]

director_name                                Andrew Stanton
num_critic_for_reviews                                421.0
gross                                            223.806889
genres                    Adventure|Animation|Family|Sci-Fi
actor_1_name                              John Ratzenberger
movie_title                                         WALL·E 
num_voted_users                                      718837
num_user_for_reviews                                 1043.0
language                                            English
budget                                                180.0
title_year                                           2008.0
imdb_score                                              8.4
movie_facebook_likes                                  16000
profit                                            43.806889
Name: 56, dtype: object